scalyr_agent/configuration.py

```diff
@@ -367,12 +372,118 @@ def parse(self):
 
             self.__monitor_configs = list(self.__config.get_json_array("monitors"))
 
-            self.__api_keys_configs = list(self.__config.get_json_array("api_keys"))
+            # NOTE do this verifications only after all config fragments were added into configurations.
+            self.__verify_api_keys()
+            self.__verify_and_match_api_keys_in_logs()
+
+            self.__api_key_configs = list(self.__config.get_json_array("api_keys"))
 
         except BadConfiguration as e:
             self.__last_error = e
             raise e
 
+    def __verify_api_keys(self):
+        """
+        Verify all api key config entries from the "api_keys" list in config.
+        """
+        api_keys = self.__config.get_json_array("api_keys")
+        default_key_config = None
+        for key_config in api_keys:
+            api_key = key_config.get("api_key", none_if_missing=True)
+            # if the 'api_key' from this entry equals to the 'main' api key,
+            # than this api key entry is considered as default.
+            if api_key is None or api_key == self.api_key:
+                if default_key_config is None:
```

Comment:
As discussed, we don't want to penalize the user if they are re-using the same api key.  HOWEVER, this code here does raise an interesting case.  If the user is trying to assign the "default" by just having it match the by api key, then there is ambiguity if there are multiple api keys with the same value.

1.  Loop through, finding first entry that doesn't have an `api_key`.  Assign that to be `default.`  If there is more than one entry missing `api_key`, throw an error.
2. If we didn't find a default key that way, then loop through, finding first entry whose `api_key` matches the main `api_key`.  Assign that to `default`.  If there is more than one, then throw an error.

What I'm trying to accomplish here is that we do not throw an error if the same `api_key` is used multiple times as long as only one entry does not contain the `api_key` field.

Let me know if that does not make sense.

Reply:
Yes, it totally does. But I think I have one option to suggest. What if we do allow users to re-use the api key but, instead of that, we do not allow leave the 'api_key' field empty? (In other words, make the  'api_key' field mandatory/required) I think it should make things easier.

Here are the cases:

1) If there are no api key entries  with the default api key, then the default api key entry is created automatically, based on another default options(such as `default_workers_per_api_key`)
2) If the user adds one(first) entry with a default api key, then it is considered as he wants to "override" the default values from the first case.
3) The second(and subsequent) api key entry with the default api key, as we discussed, is to "fine-tune" the load on workers.

Reply:
What if we just make this simpler and require both `api_key` and `id`.  We only look at `id == "default"` to say this is the default entry.  We throw an error if that entry's `api_key` does not match the main api key's value.